RHSSO Helper Refactor

ColeHiggins2 opened the pull request:
Draft work for RHSSO Helper Refactor

mshriver commented on robottelo/utils/sso.py:
```diff
@@ -0,0 +1,218 @@
+
+class SSOHost(Host):
+    def __init__(self, **kwargs):
+        self.hostname = kwargs.get('hostname', settings.rhsso.hostname)
```
docblock here please to explain why you're injecting settings.rhsso and what you're expecting to read from it.

mshriver commented on robottelo/utils/sso.py:
```diff
@@ -0,0 +1,218 @@
+            '--server {}/auth '
+            '--realm {} '
+            '--user {} '
+            '--password {}'.format(
```
f-strings throughout please

JacobCallahan reviewed (changes requested):
Just some small suggestions. Really great addition!

latran commented:
@ColeHiggins2 Any updates on this from the changes suggested?

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

mshriver commented on robottelo/utils/sso.py:
```diff
@@ -0,0 +1,197 @@
+from robottelo.datafactory import valid_emails_list
+
+
+class SSOHost(Host):
```
@JacobCallahan I'm starting to think `robottelo.hosts` should be a package, not a module.

JacobCallahan replied:
oooff that's something we can think about later...

mshriver commented on robottelo/utils/sso.py:
```diff
@@ -0,0 +1,197 @@
+        self.update_client_configuration(client_config)
+
+
+sso_host = SSOHost()
```
This instance needs to be provided by a fixture, not a module defined globally importable one.

mshriver commented on robottelo/utils/sso.py:
```diff
@@ -0,0 +1,197 @@
+        client_config = {
+            "redirectUris": [
+                "urn:ietf:wg:oauth:2.0:oob",
+                f"https://{settings.server.hostname}/users/extlogin/redirect_uri",
```
This should reference `self.satellite` - but requires basing the class off `ContentHost` to have access.

otherwise, any functions dealing with Satellite need to take a required `sat` kwarg.

mshriver commented on tests/foreman/destructive/test_ldap_authentication.py:
```diff
@@ -33,10 +33,10 @@
 from robottelo.rhsso_utils import run_command
 from robottelo.rhsso_utils import update_rhsso_user
 from robottelo.utils.issue_handlers import is_open
+from robottelo.utils.sso import sso_host
```
When this instance is provided via a fixture, the target sat (at the relevant scope) should be passed in.

```
@pytest.fixture
def sso_host(target_sat):
    return SSOHost(satellite=target_sat)
```

mshriver commented on robottelo/utils/sso.py:
```diff
@@ -0,0 +1,197 @@
+        """create new user in RHSSO instance and set the password"""
+        if not username:
+            username = gen_string('alphanumeric')
+        RHSSO_NEW_USER['username'] = username
```
This pattern is copied from the existing module, and should not be continued if you're here to improve the helpers state.

This constant is imported from the module, and then mutated - constants should be immutable (at least treated as such).

Consider creating a basic class, or even just a namedtuple, as an instance attribute. Then it's just tied to this one host, and can be manipulated however is necessary without impacting anything else running in parallel.

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

omkarkhatavkar commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_single_sign_on_using_rhsso

omkarkhatavkar commented:
@ColeHiggins2 can you please rebase i was able to fix many tests with rhsso

omkarkhatavkar commented on robottelo/utils/sso.py:
```diff
@@ -0,0 +1,197 @@
+        )
+        self.execute(update_cmd)
+
+    def get_oidc_token_endpoint(self):
```
we can use @property all of those getter methods and with the use of self.hostname

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

ColeHiggins2 commented:
Opening back up for review. tests will fail until https://github.com/SatelliteQE/robottelo/pull/10033 is merged

ColeHiggins2 commented:
Results with https://github.com/SatelliteQE/robottelo/pull/10033 added in

Launching pytest with arguments /home/colehiggins/projects/robottelo/tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso in /home/colehiggins/projects/robottelo

============================= test session starts ==============================

collecting ... collected 1 item

tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso 

=============================== warnings summary ===============================
../robo38/lib64/python3.8/site-packages/_pytest/config/__init__.py:1252
  /home/colehiggins/projects/robo38/lib64/python3.8/site-packages/_pytest/config/__init__.py:1252: PytestConfigWarning: Unknown config option: rp_ignore_errors
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================== 1 passed, 1 warning in 1012.34s (0:16:52) ===================

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

jyejare commented on robottelo/utils/sso.py:
```diff
@@ -0,0 +1,206 @@
+RHSSO_RESET_PASSWORD = Box(temporary="false", type="password", value="")
+
+
+class SSOHost(Host):
```
According to our helpers refactor history and best practices, , the utils module is for external tools that helps pytest/robottelo to execute the tests or which are not related to host objects or not satellite endpoint objects etc. 

So If the helpers are associated with host objects as we are having here then the best place for them is `robottelo/hosts.py`. Unfortunately `robottelo/hosts.py` is already a big file containing Chost, Sat and Capsule classes and its methodsssssss! 

So we should just initialise the SSOHost in that file and you should move all SSOHost methods to `SSOHost_mixins.py` categorised in classes and then inherit those classes/methods in SSOHost class. Refer https://github.com/SatelliteQE/robottelo/blob/master/robottelo/hosts.py#L123 where ContentHost class importing ContentHost mixins class. Also I agree with @mshriver comment of inheriting ContentHost class than Host class.

@mshriver @JacobCallahan  We should move all methods of CHost, Sat and Capsule classes methods to the mixins and just initialise classes in hosts.py. Its a separate effort and not blocking this PR for it!

omkarkhatavkar replied:
`class SSOHost` without 'Satellite' class has no value having the composition of Satellite class with SSOHost class make sense and not need to pass the satellite object in every function. As @mshriver said it is good to have self.satellite attribute associated with rhsso host class

ColeHiggins2 replied:
Fixes made. After speaking with jake, We decided to keep all contents of ssohost() in hosts.py and not break it up into a mixins file.

jyejare commented on robottelo/utils/sso.py:
```diff
@@ -0,0 +1,206 @@
+        """Getter method for fetching the client id and can be used other functions"""
+        client_name = f'{sat_obj.hostname}-foreman-openidc'
+        self.execute(
+            '{} config credentials '
```
Use f string ?

jyejare commented on robottelo/utils/sso.py:
```diff
@@ -0,0 +1,206 @@
+        """Helper method to create the RH-SSO Client Mapper"""
+        self.upload_rhsso_entity(json_content, "mapper_file")
+        self.execute(
+            "{} create clients/{}/protocol-mappers/models -r {} -f {}".format(
```
f-string !

jyejare commented on tests/foreman/destructive/test_ldap_authentication.py:
```diff
@@ -319,8 +313,13 @@ def test_external_logout_rhsso(enable_external_auth_rhsso, rhsso_setting_setup,
 
 
 def test_session_expire_rhsso_idle_timeout(
+<<<<<<< HEAD
     enable_external_auth_rhsso,
     rhsso_setting_setup,
+=======
+    rhsso_setting_setup,
+    enable_external_auth_rhsso,
+>>>>>>> updated tests to use default_sso_host
```
Looks like accidentally you pushed conflicted code. Please rectify.

jyejare reviewed (changes requested):
@ColeHiggins2 Overall all looks good with few advices.

1. Use `robottelo/hosts.py` for initialising SSOHost class. Check inline comment for more details.
2. Remove `robottelo/rhsso_utils.py` module in this PR itself.
3. Use f-strings
4. Fix conflict strings appeared in your code.

Optional:
1. Looks like we are using so much of SSO CLI command to execute, should we put efforts in using `python SSO API client` instead. Not blocking this PR for it but a future advice for component owner !

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

omkarkhatavkar commented on robottelo/utils/sso.py:
```diff
@@ -0,0 +1,190 @@
+            f"{settings.rhsso.realm}/protocol/openid-connect/auth"
+        )
+
+    def two_factor_token_rh_sso_url(self, sat_obj):
```
change this to `get_two_factor_token_rh_sso_url `

ColeHiggins2 commented:
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

jyejare commented:
@ColeHiggins2 Please resolve conflicts !

jyejare commented on robottelo/hosts.py:
```diff
@@ -1753,3 +1763,169 @@ def register_host_custom_repo(self, module_org, rhel_contenthost, repo_urls):
         rhel_contenthost.execute('subscription-manager repos --list')
+
+
+class SSOHost(Satellite):
```
Add a docstring got this host type !

jyejare commented on robottelo/hosts.py:
```diff
@@ -1753,3 +1763,169 @@ def register_host_custom_repo(self, module_org, rhel_contenthost, repo_urls):
+
+    def get_rhsso_client_id(self, sat_obj):
+        """getter method for fetching the client id and can be used other functions"""
+        client_name = f'{sat_obj.hostname}-foreman-openidc'
```
I dont see this variable is used at all, please remove it !

jyejare replied:
Also sat_obect is also not used anywhere. TBH I am not sure why we need sat_obj as a parameter. Isnt the SSOHost inheriting the Satellite class could be used as `self.hostname` to identify that satellite object.

omkarkhatavkar replied:
I personally think SSOHost should not inherit the Satellite Class, It would rather have the Satellite object as part of the composition rather than inheriting.

jyejare commented on robottelo/hosts.py:
```diff
@@ -1753,3 +1763,169 @@ def register_host_custom_repo(self, module_org, rhel_contenthost, repo_urls):
+        self.update_client_configuration(client_config, sat_obj)
+
+
+sso_host = SSOHost()
```
You dont need to initialize SSOHost here, it should be initialized while in use because one can pass the hostname to SSOHost.

omkarkhatavkar replied:
Use above line in the fixture as Mike mentioned earlier

jyejare commented on robottelo/hosts.py:
```diff
@@ -1753,3 +1763,169 @@ def register_host_custom_repo(self, module_org, rhel_contenthost, repo_urls):
+            ]
+        }
+        self.update_client_configuration(client_config, sat_obj)
+
```
TBH I really didnt like how many helpers we have added this in one file, instead this should have been externalized to in Mixins to track them separately.

@JacobCallahan !

JacobCallahan replied:
I believe it was previously requested for @ColeHiggins2 to move this class into hosts.py
With that in mind, I don't think there is enough in the class definition to justify externalizing to a mixin.

jyejare reviewed (changes requested):
ACK pending the comments !

ColeHiggins2 commented:
Work done in the last push:

-Removed sso_host = SSOHost() from robottelo/hosts.py
-updated default_sso_host fixture in Satellite_auth.py 
-Removed sat_obj as an argument and replaced all uses with self.satellite
-removed module_target_sat from any default_sso_host.function() call in the corresponding test files

jyejare commented:
@ColeHiggins2 You need to resolve the conflicts !

jyejare reviewed (approved):
ACK with instructions to rebase the code before merge !

jyejare commented:
@ColeHiggins2 Please rebase the code and resolve the conflicts, before we merge the code !

jyejare commented:
@ColeHiggins2 Please rebase the PR as we are going to merge it !

ColeHiggins2 commented:
Launching pytest with arguments /home/colehiggins/projects/robottelo/tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist in /home/colehiggins/projects/robottelo

============================= test session starts ==============================
platform linux -- Python 3.8.2, pytest-7.1.3, pluggy-0.13.1 -- /home/colehiggins/projects/robo38/bin/python3.8
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/colehiggins/projects/robottelo, configfile: pyproject.toml
plugins: reportportal-5.1.2, xdist-2.5.0, ibutsu-2.2.4, forked-1.3.0, mock-3.10.0, services-2.2.1
collecting ... collected 1 item

tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist 

=============================== warnings summary ===============================
.
.
.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================== 1 passed, 2 warnings in 681.16s (0:11:21) ===================

JacobCallahan reviewed (approved):
ACK, great work!!

omkarkhatavkar commented on tests/foreman/destructive/test_ldap_authentication.py:
```diff
@@ -411,7 +407,9 @@ def test_login_failure_rhsso_user_if_internal_user_exist(
     ).create()
-    external_rhsso_user = create_new_rhsso_user(client_id, username=username)
+    external_rhsso_user = default_sso_host.create_new_rhsso_user(
+        default_sso_host.get_rhsso_client_id(), username=username
```
just remove this `default_sso_host.get_rhsso_client_id()` as not required here, and move that as first line to get the latest token

omkarkhatavkar reviewed (approved):
ACK, pending one comment to resolve
